drivers/serial/uart_stm32.c

```diff
@@ -1389,6 +1392,88 @@ static int uart_stm32_init(const struct device *dev)
 #endif
 }
 
+#ifdef CONFIG_PM_DEVICE
+static int uart_stm32_set_power_state(const struct device *dev,
+					      uint32_t new_state)
+{
+	char tmp;
+	struct uart_stm32_data *data = DEV_DATA(dev);
+	USART_TypeDef *UartInstance = UART_STRUCT(dev);
+
+	if ((new_state == DEVICE_PM_ACTIVE_STATE) &&
+		(new_state != data->pm_state)) {
+		/* nothing special to do on UartInstance */
+		data->pm_state = new_state;
+	} else {
+		__ASSERT_NO_MSG(new_state == DEVICE_PM_LOW_POWER_STATE ||
+			new_state == DEVICE_PM_SUSPEND_STATE ||
+			new_state == DEVICE_PM_OFF_STATE);
+		if (data->pm_state == DEVICE_PM_ACTIVE_STATE) {
+			/* Make sure that no USART transfer is on-going */
+			while (LL_USART_IsActiveFlag_BUSY(UartInstance) == 1) {
+			}
+			while (LL_USART_IsActiveFlag_TC(UartInstance) == 0) {
+			}
+			/* prepare USART for Stop mode */
+			while (LL_USART_IsActiveFlag_RXNE(UartInstance)) {
+				/*
+				 * Empty RX Fifo :
+				 * RXNE flag is cleared by reading of RDR
+				 * Otherwise, characters already present in FIFO
+				 * will lead to immediate wake up
+				 */
+				tmp = LL_USART_ReceiveData8(UartInstance);
+			}
+			/* Make sure that USART is ready to receive */
+			while (LL_USART_IsActiveFlag_REACK(UartInstance) == 0) {
+			}
+			/* Clear OVERRUN flag */
+			LL_USART_ClearFlag_ORE(UartInstance);
+			/* Leave UartInstance unchanged */
+			data->pm_state = new_state;
+		}
+	}
+
+	return 0;
+}
+
+/**
+ * @brief disable the UART channel
+ *
+ * This routine is called to put the device in low power mode.
+ *
+ * @param dev UART device struct
+ *
+ * @return 0
+ */
+static int uart_stm32_pm_control(const struct device *dev,
+					 uint32_t ctrl_command,
+					 void *context, device_pm_cb cb,
+					 void *arg)
+{
+	struct uart_stm32_data *data = DEV_DATA(dev);
+
+	if (ctrl_command == DEVICE_PM_SET_POWER_STATE) {
+		uint32_t new_state = *((const uint32_t *)context);
+
+		if (new_state != data->pm_state) {
+			uart_stm32_set_power_state(dev,
+				new_state);
+			data->pm_state = new_state;
+		}
+	} else {
+		__ASSERT_NO_MSG(ctrl_command == DEVICE_PM_GET_POWER_STATE);
```

Comment:
It may be part of an `} else if () {`, then just return for the rest. If any `ctrl_command` gets added, this code could still be valid but would fail due to this assertion.

Reply:
simplification in  testing only new_state != DEVICE_PM_ACTIVE_STATE